tests/tot_dist_array.cpp

```diff
@@ -0,0 +1,1242 @@
+    });
+  }
+
+  /* The majority of the unit tests can be written in such a way that all they
+   * need is the tiled range and the resulting tensor. The tiled range is more
+   * or less the correct answer, since the ToTArrayFixture maps it to the values
+   * of the array, and the tensor is the thing to test. This function creates a
+   * std::vector of tiled range,tensor pairs spanning:
+   *
+   * - vector of vectors
+   * - vector of matrices
+   * - matrix of vectors
+   * - matrix of matrices
+   *
+   * The unit tests simply loop over the vector testing for all these scenarios
+   * with a few lines of code.
+   */
+  template<typename TupleElementType>
+  auto run_all() {
+      using trange_type = trange_type<TupleElementType>;
+      using tensor_type = tensor_type<TupleElementType>;
+    std::vector<std::pair<trange_type, tensor_type>> rv;
+
+    // Make vector of vector
+    for(auto tr: vector_tiled_ranges<TupleElementType>())
+      rv.push_back(std::make_pair(tr, tensor_of_vector<TupleElementType>(tr)));
+
+    // Make vector of matrix
+    for(auto tr: vector_tiled_ranges<TupleElementType>())
+      rv.push_back(std::make_pair(tr, tensor_of_matrix<TupleElementType>(tr)));
+
+    // Make matrix of vector
+    for(auto tr: matrix_tiled_ranges<TupleElementType>())
+      rv.push_back(std::make_pair(tr, tensor_of_vector<TupleElementType>(tr)));
+
+    // Make matrix of matrix
+    for(auto tr: matrix_tiled_ranges<TupleElementType>())
+      rv.push_back(std::make_pair(tr, tensor_of_matrix<TupleElementType>(tr)));
+
+    return rv;
+  }
+
+  // The world to use for the test suite
+  madness::World& m_world;
+
+}; // TotArrayFixture
+
+/* This function tests for exact equality between DistArray instances. By exact
+ * equality we mean:
+ * - Same type
+ * - Either both are initialized or both are not initialized
+ * - Same MPI context
+ * - Same shape
+ * - Same distribution
+ * - Same tiling
+ * - Components are bit-wise equal (i.e., 3.1400000000 != 3.1400000001)
+ *
+ * TODO: pmap comparisons
+ */
+template<typename LHSTileType, typename LHSPolicy,
+         typename RHSTileType, typename RHSPolicy>
+bool are_equal(const DistArray<LHSTileType, LHSPolicy>& lhs,
+               const DistArray<RHSTileType, RHSPolicy>& rhs) {
+  // Same type
+  if(!std::is_same_v<decltype(lhs), decltype(rhs)>) return false;
+
+  // Are initialized?
+  if(lhs.is_initialized() != rhs.is_initialized()) return false;
+  if(!lhs.is_initialized()) return true; // both are default constructed
+
+  // Same world instance?
+  if(&lhs.world() != &rhs.world()) return false;
+
+  // Same shape?
+  if(lhs.shape() != rhs.shape()) return false;
+
+  // Same pmap?
+  //if(*lhs.pmap() != *rhs.pmap()) return false;
+
+  // Same tiling?
+  if(lhs.trange() != rhs.trange()) return false;
+
+  // Same components? Here we make all ranks check all tiles
+  bool are_same = true;
+  for(auto idx : lhs.range()){
+    const auto& lhs_tot = lhs.find(idx).get();
```

Comment:
@evaleev I'm getting thread deadlock here when line 874 calls `are_equal`. Any insight?

Reply:
I wasn't setting the tile so this has been resolved.